internal/controllers/machineset/machineset_controller.go

```diff
@@ -178,9 +184,15 @@ func (r *Reconciler) Reconcile(ctx context.Context, req ctrl.Request) (_ ctrl.Re
 		}
 	}()
 
-	// Ignore deleted MachineSets, this can happen when foregroundDeletion
-	// is enabled
+	// Handle deletion reconciliation loop.
 	if !machineSet.DeletionTimestamp.IsZero() {
+		return r.reconcileDelete(ctx, machineSet)
+	}
+
+	// Add finalizer first if not set to avoid the race condition between init and delete.
+	// Note: Finalizers in general can only be added when the deletionTimestamp is not set.
+	if !controllerutil.ContainsFinalizer(machineSet, clusterv1.MachineSetFinalizer) {
+		controllerutil.AddFinalizer(machineSet, clusterv1.MachineSetFinalizer)
 		return ctrl.Result{}, nil
```

Comment:
Same as above, should we requeue immediately?

Reply:
my two cents: I think it makes sense to let the update event that actually persists the finalizer to trigger the next reconciliation loop as otherwise the finalizer is not guaranteed to exist and we might end up in the same code path (in practice it wouldn't make a big difference though).

Reply:
I agree with @enxebre 
@vincepri .. why should we reque explicitly here... update event would make it reque afteralll...

Reply:
Yup, no need to set Requeue